File: Study5 - Tree/Week32/Day218/boj19542_hsk.js
//  nodejs로 stack size exceeded 문제로 실패
const fs = require('fs');
const stdin = (
  process.platform === 'linux'
    ? fs.readFileSync('/dev/stdin').toString().trim()
    : `6 1 1
1 2
2 3
2 4
3 5
5 6`
).split('\n');

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const calcMaxDepthAbleToDeliver = (curNode) => {
  let maxDepth = 0;

  for (let nextNode of tree[curNode]) {
    if (visitedNodes[nextNode]) continue;
    visitedNodes[nextNode] = true;

    maxDepth = Math.max(maxDepth, calcMaxDepthAbleToDeliver(nextNode) + 1);
  }

  maxDepthOfNodeList[curNode] = maxDepth;
  return maxDepth;
};

const getTotalDistanceToMove = () => {
  let distance = 0;

  for (let i = 1; i <= N; i++) {
    if (maxDepthOfNodeList[i] < D || i === S) continue;

    distance += 1;
  }

  return distance * 2;
};

const [N, S, D] = input().split(' ').map(Number);
const tree = Array.from(new Array(N + 1), () => new Array());
const visitedNodes = new Array(N + 1).fill(false);
const maxDepthOfNodeList = new Array(N + 1).fill(0);

for (let i = 1; i < N; i++) {
  const [nodeA, nodeB] = input().split(' ').map(Number);

  tree[nodeA].push(nodeB);
  tree[nodeB].push(nodeA);
}

visitedNodes[S] = true;
calcMaxDepthAbleToDeliver(S);
console.log(getTotalDistanceToMove());

File: Study5 - Tree/Week32/Day218/boj19542_hsk.py
# nodejs로 stack size exceeded 문제로 실패하여 python으로 풀이
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10 ** 9)


def calcMaxDepthAbleToDeliver(curNode):
    maxDepth = 0

    for nextNode in tree[curNode]:
        if (visitedNodes[nextNode]): continue
        visitedNodes[nextNode] = True
        
        maxDepth = max(maxDepth, calcMaxDepthAbleToDeliver(nextNode) + 1)

    maxDepthOfNodeList[curNode] = maxDepth
    return maxDepth

def getTotalDistanceToMove():
    distance = 0
    
    for i in range(1, N + 1):
        if maxDepthOfNodeList[i] < D or i == S: continue;        
        distance += 1;
    
    return distance * 2;

N, S, D = map(int, input().split())
tree = [[] for _ in range(N + 1)]
## 항상 파이썬 문법 배워가요
maxDepthOfNodeList = [0] * (N + 1)
visitedNodes = [False] * (N + 1)

for _ in range(N - 1):
    nodeA, nodeB = map(int, input().split())
    tree[nodeA].append(nodeB)
    tree[nodeB].append(nodeA)

## 세미콜론 있어도 괜찮아요?
visitedNodes[S] = True;
calcMaxDepthAbleToDeliver(S);
print(getTotalDistanceToMove())